### PR TITLE
Add `bowtie perf` command

### PR DESCRIPTION
**Pull request by sudo-jarvis**
----
📚 Documentation preview 📚: https://bowtie-json-schema--1370.org.readthedocs.build/en/1370/

### CODE REVIEW

File: bowtie/_cli.py [hunk 1]
```diff
@@ -1198,6 +1203,14 @@ def convert(
 @SET_SCHEMA
 @TIMEOUT
 @VALIDATE
+@click.option(
```
Comment (Julian) [inline bowtie/_cli.py:1206]: If this isn't meant to be exposed, better to have it be an environment variable we check probably, which is a bit more common.

File: bowtie/_cli.py [hunk 2]
```diff
@@ -1271,6 +1298,179 @@ def validate(
...
+@dialect_option()
+@format_option()
+@click.option(
+    "--runs",
+    "-r",
+    "runs",
+    type=int,
+    default=3,
+    show_default=True,
+    help=("Number of runs used to run benchmarks."),
```
Comment (Julian) [inline bowtie/_cli.py:1312]: Parens aren't needed in all these, they're just needed if you want multiple lines.

File: bowtie/_cli.py [hunk 3]
```diff
@@ -1713,6 +1918,11 @@ async def _run(
             STDERR.print("[bold red]No test cases ran.[/]")
         elif count > 1:  # XXX: Ugh, this should be removed when Reporter dies
             STDERR.print(f"Ran [green]{count}[/] test cases.")
+
+        if time_measurement_path:
+            with time_measurement_path.open("a") as file:
```
Comment (Julian) [inline bowtie/_cli.py:1923]: Use `Path.write_text`.

Comment (Julian) [reply bowtie/_cli.py:1923]: Oh. You're appending here?

File: bowtie/_core.py [hunk 4]
```diff
@@ -859,3 +860,31 @@ def without_expected_results(self) -> Message:
 def registry():
     resources = referencing_loaders.from_traversable(files("bowtie.schemas"))
     return EMPTY_REGISTRY.with_resources(resources).crawl()
+
+
+def convert_table_to_markdown(
```
Comment (Julian) [inline bowtie/_core.py:865]: Can you have another look to make sure this function really doesn't exist in some markdown library? It seems a bit absurd for it not to.

File: bowtie/benchmarks/contains.py [hunk 5]
```diff
@@ -0,0 +1,21 @@
+def get_benchmark():
```
Comment (Julian) [inline bowtie/benchmarks/contains.py:1]: Might want to split off a separate PR to add this `bowtie/benchmarks` directory with all the benchmarks, that might be easier to immediately add? Up to you though.

File: bowtie/benchmarks/draft2020_metaschema.json [hunk 6]
```diff
@@ -0,0 +1,163 @@
+{
```
[the remark this answers is not in the document]
Comment (sudo-jarvis) [reply bowtie/benchmarks/draft2020_metaschema.json:1]: Yeah actually I felt if we have static files then atleast we would know that we just need to re run the benchmarks during deployment only when there is some change in the benchmarks folder

File: tests/test_benchmarks.py [hunk 7]
```diff
@@ -0,0 +1,433 @@
...
+    async def test_filtering_by_name(self):
+        stdout, stderr = await bowtie(
+            "filter-benchmarks",
+            "-t",
+            "keyword",
+            "-n",
+            uuid.uuid4().hex,
+            exit_code=0,
+        )
+        assert stdout == "", stderr == ""
```
Comment (Julian) [inline tests/test_benchmarks.py:433]: This assert probably doesn't do what you expect, the syntax is `assert foo, message` where the latter is a message shown on failure (so here you'd show a message of `True` or `False` but not assert against it). Are you trying to assert 2 things here presumably?

File: bowtie/_cli.py (hunk 3 above)
Comment (sudo-jarvis) [reply bowtie/_cli.py:1923]: Yeah because the same benchmark would run the same test multiple times to avg out, we would need all that value linked to one benchmark only. So one file per benchmark test.

File: tests/test_benchmarks.py (hunk 7 above)
Comment (sudo-jarvis) [reply tests/test_benchmarks.py:433]: Oh my bad, I thought it did the same

File: bowtie/_cli.py (hunk 1 above)
Comment (sudo-jarvis) [reply bowtie/_cli.py:1206]: I am not sure how that would happen, then we would have to somehow set the enivornment variable from inside bowtie itself before running the pyperf benchmark. But idts the env var would persist when pyperf then runs its own process and tries running `bowtie run`. So when bowtie run would then check whether the env var exists or not, it would always say it doesnt exist, no ? @Julian

File: tests/test_benchmarks.py [hunk 8]
```diff
@@ -0,0 +1,433 @@
...
+bowtie_dir = Path(__file__).parent.parent.joinpath("bowtie")
+default_benchmarks_dir = bowtie_dir.joinpath("benchmarks")
+keyword_benchmarks_dir = bowtie_dir.joinpath("benchmarks").joinpath("keywords")
+
+DIRECT_CONNECTABLE = "python-jsonschema"
+
+
+@pytest.fixture()
+def valid_single_benchmark():
+    module_name = "tests.benchmarks"
```
Comment (Julian) [inline tests/test_benchmarks.py:42]: There's nothing dynamic here, why avoid the `import` statement?

File: tests/test_benchmarks.py [hunk 9]
```diff
@@ -0,0 +1,433 @@
...
+            module="bowtie.benchmarks",
+        )
+        for benchmark_group in benchmark_groups:
+            assert benchmark_validated(benchmark_group.serializable())
+
+
+class TestBenchmarker:
+
+    def test_default_benchmarker(self, benchmarker_run_args):
+        _benchmarks.Benchmarker.from_default_benchmarks(**benchmarker_run_args)
```
Comment (Julian) [inline tests/test_benchmarks.py:219]: Is this meant to test it doesn't blow up with those arguments, or what?

File: bowtie/_core.py (hunk 4 above)
Comment (Julian) [reply bowtie/_core.py:865]: OK, fair enough then, thanks!

File: bowtie/benchmarks/draft2020_metaschema.json (hunk 6 above)
[the remark this answers is not in the document]
Comment (Julian) [reply bowtie/benchmarks/draft2020_metaschema.json:1]: Can you elaborate on what you're thinking there -- when/why would we run benchmarks during a deployment?
The things we'd run during a deployment are generally meant to ensure the deployment works correctly -- the only real "exception" is triggering a rebuild of the reports -- but that we do because it's 1) reasonably fast and 2) tests that the images we just rebuilt function.

File: tests/test_benchmarks.py (hunk 9 above)
Comment (sudo-jarvis) [reply tests/test_benchmarks.py:219]: Yeah just testing if the benchmarker loads up fine. So that in case something blows up we know where its blowing up.

File: tests/test_benchmarks.py (hunk 8 above)
Comment (sudo-jarvis) [reply tests/test_benchmarks.py:42]: Ended up following the same method here too. You are right, it isnt needed here.

File: bowtie/benchmarks/draft2020_metaschema.json (hunk 6 above)
[the remark this answers is not in the document]
Comment (sudo-jarvis) [reply bowtie/benchmarks/draft2020_metaschema.json:1]: My bad, I was thinking that we would need to recreate the benchmark artifacts each time we deploy. But ig we can simply run the benchmarks say weekly and during deployment it would just pick the latest artifact from the last successful run right ?

File: bowtie/benchmarks/keywords/draft2020-12/additionalProperties.py [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+                        description="Valid",
+                        instance=_format_properties_as_instance(valid),
+                    ),
+                ],
+            )
+        )
+
+        array_size *= 10
+
+    return dict(
```
[the remark this answers is not in the document]
Comment (sudo-jarvis) [reply bowtie/benchmarks/keywords/draft2020-12/additionalProperties.py:69]: Hmm that can be done, I already have such a type  - `BenchmarkGroup`. We just create it after loading the data instead of directly doing that in the benchmark file itself. Any particular advantage that it would have over the current thing ?

File: bowtie/_cli.py (hunk 1 above)
Comment (sudo-jarvis) [reply bowtie/_cli.py:1206]: I tried doing this, the env variable didnt persist. Any other idea for the same?

File: bowtie/benchmarks/draft2020_metaschema.json (hunk 6 above)
[the remark this answers is not in the document]
Comment (Julian) [reply bowtie/benchmarks/draft2020_metaschema.json]: Right yeah that was what I was thinking!

File: bowtie/_cli.py (hunk 1 above)
Comment (sudo-jarvis) [reply bowtie/_cli.py:1206]: Yeahh @Julian , ig when pyperf creates a process of its own, it doesnt inherit the env variables as also mentioned in : https://github.com/psf/pyperf/issues/47

Comment (sudo-jarvis) [reply bowtie/_cli.py:1206]: Oh wait, they did introduce some flag later on to enable inherting them!

Comment (sudo-jarvis) [reply bowtie/_cli.py]: Works now @Julian.

File: bowtie/_core.py [hunk 11]
```diff
@@ -714,12 +714,13 @@ def syntax(self) -> RenderableType:
     @classmethod
     def from_dict(
         cls,
+        instance: Any = None,
```
Comment (Julian) [inline bowtie/_core.py:717]: I'm not a fan of this -- I assume you mean to make this optional (by `None`)? But that's extra icky because `None` is a perfectly valid instance. I'll try to see where you're relying on this but it would seem to suggest another type would be better for whatever you're using this for.

File: tests/benchmarks/invalid_benchmark.py [hunk 12]
```diff
@@ -0,0 +1,2 @@
+def get_benchmark():
```
Comment (Julian) [inline tests/benchmarks/invalid_benchmark.py:1]: The tests are inside the package now as of a week or two ago, so I'm not sure these are even running? Can you move them there and/or double check.

File: tests/test_benchmarks.py [hunk 13]
```diff
@@ -0,0 +1,439 @@
...
+bowtie_dir = Path(__file__).parent.parent.joinpath("bowtie")
+default_benchmarks_dir = bowtie_dir.joinpath("benchmarks")
+keyword_benchmarks_dir = bowtie_dir.joinpath("benchmarks").joinpath("keywords")
+
+DIRECT_CONNECTABLE = "python-jsonschema"
+
+
+@pytest.fixture()
+def valid_single_benchmark():
+    return dict(
```
Comment (Julian) [inline tests/test_benchmarks.py:41]: Is this right? Isn't the expected type one of your `BenchmarkGroup` things now?

File: bowtie/_cli.py [hunk 14]
```diff
@@ -1271,6 +1277,179 @@ def validate(
...
+    quiet: bool,
+    test_suite: tuple[Iterable[TestCase], Dialect, dict[str, Any]],
+    benchmark_files: Iterable[str],
+    **kwargs: Any,
+):
+    """
+    Perform performance measurements across supported implementations.
+    """
+    try:
+        if benchmark_files:
```
Comment (Julian) [inline bowtie/_cli.py:1375]: Generally I think it's nicer in general to avoid `if/elif` chains -- specifically if you want a small exercise, here I think the pattern would be similar to what we do elsewhere, namely when you define each of these options, they shouldn't set boolean flags, they should set a `deserialize` argument (so `keywords` sets it to `Benchmarker.for_keywords` or some wrapper callable for that, `test_suite` to `from_test_cases`, etc.)

File: bowtie/_cli.py [hunk 15]
```diff
@@ -1271,6 +1277,179 @@ def validate(
...
+
+    return 0
+
+
+@subcommand
+@dialect_option()
+@click.option(
+    "-t",
+    "--benchmark-type",
+    type=click.Choice(["default", "keyword"]),
```
Comment (Julian) [inline bowtie/_cli.py:1422]: Doesn't have to be immediately (i.e. we can leave it and "fix"/"change") but I think `tags` or `markers` makes a bit more sense (i.e. that we have benchmarks have some notion of tags [a la website] or markers [a la pytest test markers]) and then you can filter by a tag/marker, and then right now we just have 1 of those, for keyword benchmarks.

File: bowtie/schemas/benchmarks.json [hunk 16]
```diff
@@ -0,0 +1,66 @@
+{
+  "title": "Bowtie Benchmarks",
+  "description": "A collection of benchmarks used to compare the performance between various JSON Schema Implementations.",
```
Comment (Julian) [inline bowtie/schemas/benchmarks.json:3]:
```suggestion
  "description": "A collection of benchmarks used to compare the performance between various JSON Schema implementations.",
```

File: docs/spelling-wordlist.txt [hunk 17]
```diff
@@ -18,6 +18,7 @@ implementers
 linter
 linters
 positionally
+perf
```
Comment (Julian) [inline docs/spelling-wordlist.txt:21]: I mean `perf` isn't really a word -- do you see where it's creeping in in the docs? Possibly somewhere should enclose it in ` (which would make this go away)?

File: tests/test_benchmarks.py (hunk 13 above)
Comment (sudo-jarvis) [reply tests/test_benchmarks.py:41]: Didnt do the same refactoring for tests properly. They are still loading from a dict. Fixing it now.

File: tests/benchmarks/invalid_benchmark.py (hunk 12 above)
Comment (sudo-jarvis) [reply tests/benchmarks/invalid_benchmark.py:1]: I am not sure I get it. Could you please elaborate it further ?

Comment (Julian) [reply tests/benchmarks/invalid_benchmark.py:1]: These files belong in `bowtie/tests/benchmarks`, right now they probably don't even run in CI at all.

Comment (sudo-jarvis) [reply tests/benchmarks/invalid_benchmark.py:1]: `bowtie/tests/test_benchmarks.py` is loading the file, so it is being used right? Or else the tests would also fail.

Comment (Julian) [reply tests/benchmarks/invalid_benchmark.py:1]: I'm looking at the GitHub UI not the git branch, so maybe you're saying the UI is wrong, but `bowtie/tests/test_benchmarks.py` doesn't exist, you have `tests/test_benchmarks.py`, which I again suspect isn't running at all, now that the tests are at `bowtie/tests`!

File: docs/spelling-wordlist.txt (hunk 17 above)
Comment (sudo-jarvis) [reply docs/spelling-wordlist.txt:21]: I havent myself included perf anywhere in the docs yet, ig when its building the docs and getting the supported commands it would be adding perf right ? That is what it raising the spelling error to handle which I added the keyword in our wordlist.

File: bowtie/_core.py (hunk 11 above)
Comment (sudo-jarvis) [reply bowtie/_core.py:717]: I am not exactly able to recall why I added this :( . It was probably the None instance failing bowtie run or something in some of the benchmark. It was raising the error instance cannot be left empty.

File: docs/spelling-wordlist.txt (hunk 17 above)
Comment (Julian) [reply docs/spelling-wordlist.txt:21]: Oh, that's odd, I guess that's a minor bug in `sphinx-click`'s output that hasn't been obvious given all the other command names are words... I mean this is pretty tiny obviously but perhaps put this instead in `docs/cli.rst` just above the click directive:

```
.. spelling:word-list::
   perf
```

so that at least it only applies in that file?

File: tests/benchmarks/invalid_benchmark.py (hunk 12 above)
Comment (sudo-jarvis) [reply tests/benchmarks/invalid_benchmark.py:1]: Oh my bad :( I didnt merge the main branch and didnt realize the location of tests has changed. Will change that

File: bowtie/_cli.py (hunk 14 above)
Comment (sudo-jarvis) [reply bowtie/_cli.py]: Hi @Julian , tried doing this. Can you please have a look ?

File: bowtie/_cli.py (hunk 15 above)
Comment (sudo-jarvis) [reply bowtie/_cli.py:1422]: I'll open an issue for this, and if time permits then would add this functionality.

File: bowtie/_cli.py (hunk 14 above)
Comment (sudo-jarvis) [reply bowtie/_cli.py:1375]: The arguements for each are different but, any idea how to handle that?

Comment (sudo-jarvis) [reply bowtie/_cli.py]: Could you please have a look now? @Julian
